systest/cluster/nodes.go

```diff
@@ -571,7 +599,108 @@ func deployNodes(ctx *testcontext.Context, kind string, from, to int, opts ...De
+	goldenAtxId types.ATXID,
+	opts ...DeploymentOpt,
+) ([]*NodeClient, error) {
+	ctx.Log.Debugw("deploying remote nodes", "from", from, "to", to)
+	var (
+		eg      errgroup.Group
+		clients = make(chan *NodeClient, to-from)
+		cfg     = SmesherDeploymentConfig{}
+	)
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	if delta := to - from; len(cfg.keys) != delta {
+		return nil, fmt.Errorf(
+			"keys must be overwritten for all members of the cluster: delta %d, keys %d %v",
+			delta,
+			len(cfg.keys),
+			cfg.keys,
+		)
+	}
+	for i := from; i < to; i++ {
+		i := i
+		finalFlags := make([]DeploymentFlag, len(cfg.flags), len(cfg.flags)+ctx.PoetSize)
+		copy(finalFlags, cfg.flags)
+		if !cfg.noDefaultPoets {
+			var poetIds []int
+			for idx := 0; idx < ctx.PoetSize; idx++ {
+				poetIds = append(poetIds, idx)
+			}
+			finalFlags = append(finalFlags, PoetEndpoints(poetIds...))
+		}
+		if ctx.BootstrapperSize > 1 {
+			finalFlags = append(finalFlags, BootstrapperUrl(BootstrapperEndpoint(i%ctx.BootstrapperSize)))
+		} else {
+			finalFlags = append(finalFlags, BootstrapperUrl(BootstrapperEndpoint(0)))
+		}
+
+		finalFlags = append(finalFlags, SmesherKey(cfg.keys[i-from]))
+
+		nodeId := fmt.Sprintf("%s-%d", smesherApp, i)
+		eg.Go(func() error {
+			labels := nodeLabels(smesherApp, nodeId)
+			labels["bucket"] = strconv.Itoa(i % buckets)
+			if err := deployNode(ctx, nodeId, labels, finalFlags); err != nil {
+				return err
+			}
+			deployNodeSvc(ctx, nodeId)
+			clients <- &NodeClient{
+				session: ctx,
+				Node: Node{
+					Name:      nodeId,
+					P2P:       7513,
+					GRPC_PUB:  9092,
+					GRPC_PRIV: 9093,
+				},
+			}
+			return nil
+		})
+
+		eg.Go(func() error {
+			postId := fmt.Sprintf("%s-%d", postServiceApp, i)
+			labels := nodeLabels(postServiceApp, postId)
+			labels["bucket"] = strconv.Itoa(i % buckets)
+			labels["nodeId"] = nodeId
+			err := deployPostService(ctx, postId, labels, nodeId,
+				hex.EncodeToString(cfg.keys[i-from].Public().(ed25519.PublicKey)),
+				goldenAtxId.Hash32().Hex()[2:],
+			)
+			return err
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	close(clients)
+	var rst []*NodeClient
+	for node := range clients {
+		rst = append(rst, node)
+	}
+	sort.Slice(rst, func(i, j int) bool {
+		return decodeOrdinal(rst[i].Name) < decodeOrdinal(rst[j].Name)
+	})
+	return rst, nil
+}
+
 func deleteNode(ctx *testcontext.Context, id string) error {
+	// find and delete any post services linked to this node
+	var errService error
+	if svcs, err := ctx.Client.AppsV1().
+		Deployments(ctx.Namespace).
+		List(ctx, apimetav1.ListOptions{LabelSelector: fmt.Sprintf("nodeId=%s", id)}); err == nil {
+		for _, svc := range svcs.Items {
+			err = ctx.Client.AppsV1().Deployments(ctx.Namespace).
+				Delete(ctx, svc.ObjectMeta.Name, apimetav1.DeleteOptions{})
+			errService = errors.Join(errService, err)
+		}
+	}
+	if errService != nil {
+		return errService
+	}
```

Comment:
You wouldn't need it if post-service was deployed as a second container on the node Pod. You don't need to have separate Deployments for the node and post-service.

Reply:
This is true in the current (simple) test setup where we only do 1:1 deployments.

I'm planning on adding more tests later where post services are dynamically added (and maybe even removed) and I think for this we need separate deployments 😕